File: day_2/python/main.py
import sys, os

def read_data(file_path: str):
    with open(file_path, "r") as f:
        data_ = f.read().strip()
    return data_.split("\n")


if __name__ == "__main__":
    input_file_path = sys.argv[1]
    abs_path = os.path.abspath(input_file_path)
    data = read_data(abs_path)
    target = {"red": 12, "green": 13, "blue": 14}
    game_dict = {}
    for x in data:
        game_id, sets = x.split(": ")
        id_ = game_id.split(" ")[-1]
        sets = sets.split("; ")
        game_dict.update({id_: {"red": 0, "green": 0, "blue": 0}})
        for y in sets:
            colors = y.split(", ")
            for z in colors:
                amount, color = z.split(" ")
                amount_int = int(amount)
                current_amount = game_dict.get(id_)
                assert current_amount is not None
                new_amount = max(current_amount.get(color), amount_int)
                current_amount.update({color: new_amount})
                game_dict.update({id_: current_amount})

    filter_1 = dict(filter(
        lambda x: (
            x[1].get("red") <= target.get("red") and
            x[1].get("green") <= target.get("green") and
            x[1].get("blue") <= target.get("blue")
        ),
        game_dict.items()
    ))
    result_1 = sum(int(x) for x in filter_1.keys())
    print(result_1)

    filter_2 = dict(map(
        lambda x: (
                x[0], x[1].get("red") * x[1].get("green") * x[1].get("blue")
        ),
        game_dict.items()
    ))
    result_2 = sum(x for x in filter_2.values())
    print(result_2)

File: day_2/go/main.go
package main

import (
    "fmt"
	"os"
	"path/filepath"
	"strings"
    "strconv"
)

func check(e error) {
    if e != nil {
        panic(e)
    }
}

type Game struct {
    id int
    red int
    green int
    blue int
}

func main() {
    path := os.Args[1]
    absPath, err := filepath.Abs(path)
    check(err)
    dat, err := os.ReadFile(absPath)
    dat_string := strings.Trim(string(dat), "\n")
    lines := strings.Split(dat_string, "\n")
    games := []Game{}
    for _, x := range lines {
        game_colors_split := strings.Split(x, ": ")
        game_id := strings.Split(game_colors_split[0], " ")[1]
        id_int, err := strconv.Atoi(game_id)
        check(err)
        colors := strings.Split(game_colors_split[1], "; ")
        new_game := Game{id_int, 0, 0, 0}
        for _, y := range colors {
            color_groups_split := strings.Split(y, ", ")
            for _, z := range color_groups_split {
                color_amount_split := strings.Split(z, " ")
                color := color_amount_split[1]
                amount, err := strconv.Atoi(color_amount_split[0])
                check(err)

                switch color {
                case "red":
                    new_game.red = max(new_game.red, amount)
                case "green":
                    new_game.green = max(new_game.green, amount)
                case "blue":
                    new_game.blue = max(new_game.blue, amount)
                }
            }
        }
        games = append(games, new_game)
    }
    target := Game{-1, 12, 13, 14}
    result_1 := 0
    for _, game := range games {
        if game.red <= target.red && game.green <= target.green && game.blue <= target.blue {
            result_1 += game.id
        }
    }
    fmt.Println(result_1)

    result_2 := 0
    for _, game := range games {
        result_2 += game.red * game.green * game.blue
    }
    fmt.Println(result_2)
}
